File: benchmarks/tsb/bench_where_mask_df_fn.ts
/**
 * Benchmark: whereDataFrame / maskDataFrame — standalone functional where/mask for DataFrame.
 * Outputs JSON: {"function": "where_mask_df_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, whereDataFrame, maskDataFrame } from "../../src/index.ts";

const ROWS = 100_000;
const WARMUP = 5;
const ITERATIONS = 20;

const df = DataFrame.fromColumns({
  a: Array.from({ length: ROWS }, (_, i) => i * 1.0),
  b: Array.from({ length: ROWS }, (_, i) => (i % 2 === 0 ? null : i * 0.5)),
  c: Array.from({ length: ROWS }, (_, i) => i * -1.0),
});

const condFn = (v: unknown) => (v as number) > 0;

for (let i = 0; i < WARMUP; i++) {
  whereDataFrame(df, condFn, { other: 0 });
  maskDataFrame(df, condFn, { other: -1 });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  whereDataFrame(df, condFn, { other: 0 });
  maskDataFrame(df, condFn, { other: -1 });
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "where_mask_df_fn",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_where_mask_df_fn.py
"""
Benchmark: pandas DataFrame.where() / DataFrame.mask() — conditional replacement.
Outputs JSON: {"function": "where_mask_df_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd
import numpy as np

ROWS = 100_000
WARMUP = 5
ITERATIONS = 20

df = pd.DataFrame({
    "a": [i * 1.0 for i in range(ROWS)],
    "b": [float("nan") if i % 2 == 0 else i * 0.5 for i in range(ROWS)],
    "c": [i * -1.0 for i in range(ROWS)],
})

cond = df > 0

for _ in range(WARMUP):
    df.where(cond, other=0)
    df.mask(cond, other=-1)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    df.where(cond, other=0)
    df.mask(cond, other=-1)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({"function": "where_mask_df_fn", "mean_ms": round(mean_ms, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))
